### PR TITLE
Merkle storage update (rebased, was #439)

### PR DESCRIPTION
**Pull request by mat13mn**

### CODE REVIEW

File: storage/src/context_action_storage.rs [hunk 1]
```diff
@@ -89,7 +89,7 @@ impl ContextActionStorage {
     #[inline]
     pub fn put_action(
         &mut self,
-        block_hash: &BlockHash,
+        block_hash: BlockHash,
```
Comment (tizoc) [inline storage/src/context_action_storage.rs:92]: Why is this no longer a reference?

File: crypto/src/hash.rs [hunk 2]
```diff
@@ -53,6 +54,9 @@ macro_rules! define_hash {
             std::cmp::Ord,
             std::hash::Hash,
         )]
+        // TODO: Hash -> Array, in current shape it doesnt seem to be
+        // super effective object is still allocated on heap and clone is
+        // costful, even
```
Comment (tizoc) [inline crypto/src/hash.rs:59]: Still has to be evaluated, but we have this: https://github.com/simplestaking/tezedge/pull/424

File: shell/src/context_listener.rs [hunk 3]
```diff
@@ -3,21 +3,31 @@
 
 //! Listens for events from the `protocol_runner`.
 
+use bytes::{Buf, BufMut, BytesMut};
```
Comment (tizoc) [inline shell/src/context_listener.rs:6]: There are many clippy warnings regarding unused and duplicated imports on this file

File: light_node/src/configuration.rs [hunk 4]
```diff
@@ -850,11 +858,21 @@ impl Environment {
...
-                        .unwrap_or("true")
-                        .parse::<bool>()
-                        .expect("Provided value cannot be converted to bool"),
+                    action_store_backend: {
+                        let backend = args
+                            .value_of("actions-store-backend")
+                            .unwrap_or("none")
+                            .parse::<String>()
+                            .expect("Provided value cannot is not a string");
+                        match &backend[..] {
```
Comment (tizoc) [inline light_node/src/configuration.rs:867]: `backend.as_str()` instead?

File: shell/tests/apply_and_verify_context_actions_from_file.rs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+                | ContextAction::Copy { .. }
+                | ContextAction::Delete { .. }
+                | ContextAction::RemoveRecursively { .. }
+                | ContextAction::Commit { .. }
+                | ContextAction::Checkout { .. } => {
+                    if let Err(e) = perform_context_action(&msg.action, &mut context) {
+                        panic!("cannot perform action {:?} error: '{}'", &msg, e);
+                    }
+                }
+                _ => {}
```
Comment (tizoc) [inline shell/tests/apply_and_verify_context_actions_from_file.rs:184]: Please, use an explicit enumeration here, we should avoid wildcard matches on enums whenever possible.

File: storage/Cargo.toml [hunk 6]
```diff
@@ -12,21 +12,34 @@ derive_builder = "0.9"
 failure = "0.1"
 getset = "0.1"
 hex = "0.4"
+im = { version = "15.0.0", features = ["serde"] }
 itertools = "0.10"
 num_cpus = "1.13"
 rocksdb = {version = "0.15", features = ["snappy", "lz4", "zstd", "zlib"], default-features = false }
 serde = { version = "1.0", features = ["derive", "rc"] }
-slog = "2.7"
+slog = "2.5"
```
Comment (tizoc) [inline storage/Cargo.toml:20]: Version is being lowered here,  but shouldn't

File: storage/scripts/actions_tool.rs [hunk 7]
```diff
@@ -0,0 +1,257 @@
+use bytes::{Buf, BufMut, BytesMut};
+use std::error::Error;
+use std::fs::{File, OpenOptions};
+use std::io::{BufReader, Read, Seek, SeekFrom};
+use std::path::Path;
+// use bytes::buf::BufExt;
```
Comment (tizoc) [inline storage/scripts/actions_tool.rs:6]: The commented imports should be removed instead.

File: storage/scripts/actions_tool.rs [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+#[derive(Clone, Copy, Debug)]
+pub struct ActionsFileHeader {
+    pub block_height: u32,
+    pub actions_count: u32,
+    pub block_count: u32,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub enum ContextAction {
```
Comment (tizoc) [inline storage/scripts/actions_tool.rs:28]: This seems to be in the old format, right? I don't see any mention of `tree_hash` and `new_tree_hash`.

File: storage/scripts/merkle_storage_stats.rs [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+            .arg(Arg::with_name("actions_file")
+                 .required(true)
+                 .help("path to the actions.bin")
+                 .index(1))
+            .arg(Arg::with_name("backend")
+                 .short("b")
+                 .long("backend")
+                 .default_value("in-memory-gced")
+                 .help("backend to use for storing merkle storage. Possible values: in-memory-gced, in-memory-mark-sweep-gced")
+                       // + "rocksdb, sled, in-memory-gced, in-memory-mark-sweep-gced")
```
Comment (tizoc) [inline storage/scripts/merkle_storage_stats.rs:93]: What is this comment for?

File: storage/scripts/merkle_storage_stats.rs [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+            }
+
+            match &action {
+                ContextAction::Set {
+                    key,
+                    value,
+                    ignored,
+                    ..
+                } => {
+                    if !ignored {
```
Comment (tizoc) [inline storage/scripts/merkle_storage_stats.rs:173]: This looks outdated too, we don't have an `ignored` flag amymore.

File: crypto/src/hash.rs [hunk 11]
```diff
@@ -68,6 +72,10 @@ macro_rules! define_hash {
                 }
             }
 
+            pub fn to_bytes(&self) -> &[u8] {
```
Comment (tizoc) [inline crypto/src/hash.rs:75]: Since we are borrowing, this should be `as_bytes` and not `to_bytes`

**Comment by tizoc** [comment]
@binier @mat13mn @mambisi I left a bunch of comments, but there are also many clippy warnings, please address those that are related to the code each of you wrote.

Btw, I'm confused about `storage/scripts/actions_tool.rs` having it's own definitions of context actions, shouldn't it import the definitions that already exist? Those also don't include `tree_hash` and `new_tree_hash`, so it seems like things are out of sync. Is there anything am I missing?

**Comment by binier** [comment]
> Btw, I'm confused about `storage/scripts/actions_tool.rs` having it's own definitions of context actions, shouldn't it import the definitions that already exist? Those also don't include `tree_hash` and `new_tree_hash`, so it seems like things are out of sync. Is there anything am I missing?

That's just to support the old actions.bin file. That duplicate definition should be removed so that script supports the new format.

File: storage/scripts/merkle_storage_stats.rs (hunk 9 above)
Comment (binier) [reply storage/scripts/merkle_storage_stats.rs:93]: Right now that script only supports in-memory stores. It probably should support rocksdb and sled as well, but those implementations don't implement measurements for memory usage, so it won't create relevant statistics.

Probably we should just  replace it with todo comment.

File: storage/scripts/actions_tool.rs (hunk 8 above)
Comment (binier) [reply storage/scripts/actions_tool.rs:28]: Yes its for old format.

**Comment by tizoc** [comment]
@binier understood, thank you!

File: storage/scripts/merkle_storage_stats.rs (hunk 9 above)
Comment (tizoc) [reply storage/scripts/merkle_storage_stats.rs:93]: Yes, please. Add a meaningful comment or remove that one if there is nothing to say.

File: crypto/src/hash.rs (hunk 2 above)
Comment (mat13mn) [reply crypto/src/hash.rs:59]: my impression is that thread safe Arc as well as heap allocation might cause noticeable overhead...

File: crypto/src/hash.rs (hunk 11 above)
Comment (mat13mn) [reply crypto/src/hash.rs:75]: will do

**Comment by tizoc** [comment]
Closing this one, sone parts have been merged to develop already (#458 and #453) , others are being handled in #463 and #454

**Pull request closed without merging**